File: plugins/OrchidWiki/src/main/java/com/eden/orchid/wiki/model/WikiModel.java
package com.eden.orchid.wiki.model;

import com.eden.orchid.api.theme.pages.OrchidPage;
import lombok.Getter;

import javax.inject.Singleton;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Singleton
public class WikiModel {

    @Getter
    private Map<String, WikiSection> sections;

    public void initialize() {
        this.sections = new LinkedHashMap<>();
    }

    public List<OrchidPage> getAllPages() {
        List<OrchidPage> allPages = new ArrayList<>();
        for (String key : sections.keySet()) {
            allPages.add(sections.get(key).getSummaryPage());
            allPages.addAll(sections.get(key).getWikiPages());
        }

        return allPages;
    }

}

File: plugins/OrchidWiki/src/main/kotlin/com/eden/orchid/wiki/model/WikiModel.kt
package com.eden.orchid.wiki.model

import com.eden.orchid.api.generators.OrchidGenerator
import com.eden.orchid.api.theme.pages.OrchidPage
import com.eden.orchid.wiki.pages.WikiSectionsPage

class WikiModel(
    sectionsList: List<WikiSection>
) : OrchidGenerator.Model {

    var sectionsPage: WikiSectionsPage? = null

    var sections: MutableMap<String, WikiSection> = linkedMapOf(*(sectionsList.map { it.key to it }.toTypedArray()))

    override val allPages: List<OrchidPage>
        get() {
            val allPages = ArrayList<OrchidPage>()
            for (value in sections.values) {
                allPages.add(value.summaryPage)
                allPages.addAll(value.wikiPages)
                if (value.bookPage != null) {
                    allPages.add(value.bookPage!!)
                }
            }
            if (sectionsPage != null) {
                allPages.add(sectionsPage!!)
            }

            return allPages
        }

    fun getSection(sectionKey: String): WikiSection? {
        if (sectionKey.isNotBlank()) {
            return sections.getOrDefault(sectionKey, null)
        }

        return null
    }
}
